src/unstract/sdk/file_storage/fs_permanent.py

```diff
@@ -0,0 +1,63 @@
+from typing import Any, Union
+
+from unstract.sdk.exceptions import FileOperationError, FileStorageError
+from unstract.sdk.file_storage.constants import Common
+from unstract.sdk.file_storage.fs_impl import FileStorage
+from unstract.sdk.file_storage.fs_provider import FileStorageProvider
+
+
+class PermanentFileStorage(FileStorage):
+    SUPPORTED_FILE_STORAGE_TYPES = [
+        FileStorageProvider.GCS.value,
+        FileStorageProvider.S3.value,
+        FileStorageProvider.Azure.value,
+    ]
+
+    def __init__(
+        self,
+        provider: FileStorageProvider,
+        credentials: dict[str, Any] = {},
+    ):
+        if provider.value not in self.SUPPORTED_FILE_STORAGE_TYPES:
+            raise FileStorageError(
+                f"File storage provider is not supported in Permanent mode. "
+                f"Supported providers: {self.SUPPORTED_FILE_STORAGE_TYPES}"
+            )
+        if provider == FileStorageProvider.GCS:
+            super().__init__(provider, credentials)
+        else:
+            raise NotImplementedError
+
+    def copy_on_write(self, path1, path2):
+        self.fs.put(path1, path2)
+
+    def read_file(
+        self,
+        path: str,
+        mode: str,
+        encoding: str = Common.DEFAULT_ENCODING,
+        seek_position: int = 0,
+        length: int = Common.FULL,
+    ) -> Union[bytes, str]:
+        """Read the file pointed to by the file_handle.
+
+        Args:
+            path (str): Path to the file to be opened
+            mode (str): Mode in whicg the file is to be opened. Usual options
+                        include r, rb, w and wb
+            encoding (str): Encoding type like utf-8 or utf-16
+            seek_position (int): Position to start reading from
+            length (int): Number of bytes to be read. Default is full
+            file content.
+
+        Returns:
+            Union[bytes, str] - File contents in bytes/string based on the opened mode
+        """
+        try:
+            if not self.path_exists(path):
+                local_file_storage = FileStorage(provider=FileStorageProvider.Local)
+                if local_file_storage.path_exists(path):
+                    self.copy_on_write(path, path)
```

Comment:
@gaya3-zipstack All of lines 57-60 is the "copy on write" strategy. Probably we can add a dev comment regarding the same.

This function can be named just `copy` itself.

Reply:
Done.